### PR TITLE
fix(plugin-meetings): send audio controls directly to current locus in breakouts

### PR DESCRIPTION
**Pull request by chrisadubois**
# COMPLETES https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-393351

## This pull request addresses

When a host is in a breakout session and attempts to mute participants using audio controls (`mute`/`unmute`, `muteOnEntry`, `disallowUnmute`), the SDK was sending the PATCH request to the **main session locus URL** with `authorizingLocusUrl` pointing to the breakout. However, the Locus controls API does not support `authorizingLocusUrl` for audio controls — they are not wired into the cross-locus GraphQL authorization path. The server ignores the field and evaluates the request against the target locus only, resulting in **403 Forbidden** (errorCode `2403031`) because the host is not joined to the main locus directly.

## by making the following changes

- **`constants.ts`**: Added `AUDIO_CONTROL_BODY_KEYS` as a `ReadonlySet<string>` containing `audio`, `muteOnEntry`, and `disallowUnmute` — the body keys that do not support cross-locus authorization.
- **`util.ts`**: Added three static helper methods:
  - `isAudioControl(body)` — checks if all body keys are audio control keys
  - `isBreakoutLocusUrl(locusUrl, mainLocusUrl)` — detects breakout session by URL comparison
  - `getControlsRequestParams({body, locusUrl, mainLocusUrl})` — resolves `targetUrl` and `extraBody` for the controls request, routing audio controls directly to `locusUrl` and non-audio controls to `mainLocusUrl` with `authorizingLocusUrl`
- **`index.ts`**: Both `update()` and `setControls()` now delegate routing to `Util.getControlsRequestParams()` instead of inline logic.
- **`test/.../util.js`**: Added unit tests for `isAudioControl()`, `isBreakoutLocusUrl()`, and `getControlsRequestParams()`.
- **`test/.../index.js`**: Added integration tests for all breakout audio routing scenarios.

Non-breakout behavior is completely unchanged.

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

- See comments below for additional tests completed
- `setMuteAll` in breakout → sends to `locusUrl` without `authorizingLocusUrl` (attendee and PANELIST roles)
- `setMuteOnEntry` in breakout → sends to `locusUrl` without `authorizingLocusUrl`
- `setDisallowUnmute` in breakout → sends to `locusUrl` without `authorizingLocusUrl`
- `update()` with audio scope in breakout → sends to `locusUrl` without `authorizingLocusUrl`
- `update()` with non-audio scope in breakout → sends to `mainLocusUrl` with `authorizingLocusUrl`
- Mixed `update(audio, reactions)` in breakout → audio routed to `locusUrl`, reactions to `mainLocusUrl`
- All existing non-breakout tests continue to pass (28 index tests + 21 new util tests = 49 total)
- `isAudioControl()` edge cases: empty body, single audio key, mixed keys, non-audio keys
- `isBreakoutLocusUrl()` edge cases: undefined/null/empty mainLocusUrl, matching URLs
- `getControlsRequestParams()` scenarios: non-breakout, breakout audio, breakout non-audio, mixed body keys

## The GAI Coding Policy And Copyright Annotation Best Practices ##

- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [x] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [x] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [x] Github Copilot
  - [ ] Other - Please Specify
- [x] This PR is related to
  - [ ] Feature
  - [x] Defect fix
  - [ ] Tech Debt
  - [ ] Automation

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request
- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: .github/agents/code-reviewer.agent.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+---
```
Comment (chrisadubois) [inline .github/agents/code-reviewer.agent.md]: remove this file

**Comment by chrisadubois** [comment]
Additional testing evidence I did manually demonstrating the following:

1. I can mute all in a breakout without 403 with and without other participants in and out of the breakout
2. I can mute all in a normal meeting without any regressions same as 2
3. I can take other controls actions such as raiseHand, etc in and out of these as well

Additional testing I still need to do 
1. test webinar mute all and other functionalities
2. test webinar 5k mute all and other functionalities
3. 1 and 2 in a breakout session with and without participants

Screenshot 2026-04-09 at 6 36 39 PM
Screenshot 2026-04-09 at 6 35 56 PM
Screenshot 2026-04-09 at 6 35 04 PM
Screenshot 2026-04-09 at 6 34 34 PM
Screenshot 2026-04-09 at 6 33 55 PM
Screenshot 2026-04-09 at 6 33 46 PM
Screenshot 2026-04-09 at 6 33 40 PM
Screenshot 2026-04-09 at 6 33 33 PM
Screenshot 2026-04-09 at 6 33 27 PM
Screenshot 2026-04-09 at 6 33 22 PM
Screenshot 2026-04-09 at 6 33 17 PM
Screenshot 2026-04-09 at 8 34 37 PM
Screenshot 2026-04-09 at 8 34 21 PM
Screenshot 2026-04-09 at 8 34 10 PM

**Comment by chrisadubois** [comment]
I also ran the hostMuteControlsUSM, breakoutSession, webinar, and inmeetingfeatures tests locally against a linked SDK on the web client to validate for regressions, and all of those passed

File: packages/@webex/plugin-meetings/src/controls-options-manager/util.ts [hunk 2]
```diff
@@ -400,6 +401,81 @@ class Utils {
...
+    locusUrl: string;
+    mainLocusUrl?: string;
+  }): {
+    targetUrl: string;
+    extraBody: Record<string, any>;
+  } {
+    const {body, locusUrl, mainLocusUrl} = options;
+
+    if (!locusUrl) {
+      return {targetUrl: locusUrl, extraBody: {}};
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/controls-options-manager/util.ts]: this doesn't feel right - we're returning targetUrl set to a false-y value and the code that uses getControlsRequestParams() doesn't check targetUrl so will try to send a request to a false-y url

File: packages/@webex/plugin-meetings/src/controls-options-manager/index.ts [hunk 3]
```diff
@@ -172,15 +172,16 @@ export default class ControlsOptionsManager {
     });
 
     return payloads.reduce((previous, payload) => {
-      const extraBody =
-        this.mainLocusUrl && this.mainLocusUrl !== this.locusUrl
-          ? {authorizingLocusUrl: this.locusUrl}
-          : {};
+      const {targetUrl, extraBody} = Util.getControlsRequestParams({
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/controls-options-manager/index.ts]: suggestion: maybe it would be cleaner to just return the whole body instead of just the extraBody? getControlsRequestParams has all the info it needs to create the body and it would avoid the necessity for the called to be stitching the body with extraBody which we're doing now in 2 places 

actually, maybe getControlsRequestParams() should also return the full url as the targetUrl (so including the `/${CONTROLS}` bit and the method='PATCH' too - so it basically returns all the request params  (like the function name suggests) and not just a random subset of them

Comment (chrisadubois) [reply packages/@webex/plugin-meetings/src/controls-options-manager/index.ts]: Im good with this and can make this change.

File: packages/@webex/plugin-meetings/src/controls-options-manager/util.ts (hunk 2 above)
Comment (chrisadubois) [reply packages/@webex/plugin-meetings/src/controls-options-manager/util.ts]: I'm going to change to `Promise.reject(ParameterError)` in both functions rather than at the util when this.locusUrl is falsy

File: packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js [hunk 4]
```diff
@@ -799,6 +799,195 @@ describe('plugin-meetings', () => {
...
+        it('should return false when mainLocusUrl is empty string', () => {
+          assert.isFalse(ControlsOptionsUtil.isBreakoutLocusUrl('locus/breakout', ''));
+        });
+      });
+
+      describe('getControlsRequestParams()', () => {
+        const locusUrl = 'locus/breakout';
+        const mainLocusUrl = 'locus/main';
+
+        it('should return request params with undefined-based uri when locusUrl is falsy', () => {
```
Comment (chrisadubois) [inline packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js]: this test is pointless

File: packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js [hunk 5]
```diff
@@ -799,6 +799,195 @@ describe('plugin-meetings', () => {
...
+            locusUrl,
+            mainLocusUrl,
+          });
+
+          assert.equal(result.uri, 'locus/breakout/controls');
+          assert.deepEqual(result.body, {disallowUnmute: {enabled: true}});
+          assert.equal(result.method, 'PATCH');
+        });
+
+        it('should return mainLocusUrl with authorizingLocusUrl for mixed audio and non-audio in a breakout', () => {
```
Comment (chrisadubois) [inline packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js]: This test is also kind of pointless since this cannot happen, in fact we should probably guard against usage like this

**Review by marcin-bazyl — APPROVED** [review]

File: packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js (hunk 4 above)
Comment (chrisadubois) [reply packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js]: > ⚠️ AI Assisted Content 🤖

Removed in de299d5. Callers guard with `ParameterError` before invoking `getControlsRequestParams()`, making this path unreachable through the public API.

File: packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js (hunk 5 above)
Comment (chrisadubois) [reply packages/@webex/plugin-meetings/test/unit/spec/controls-options-manager/util.js]: > ⚠️ AI Assisted Content 🤖

Removed in de299d5. `update()` sends each control scope as a separate request, and `setControls()` only handles audio-related settings — mixed payloads cannot occur through the public API. Added JSDoc notes documenting the design decision.

**Comment by chrisadubois** [comment]
Additional tests with webinar and breakouts (webinar 5k)
Screenshot 2026-04-10 at 9 36 57 AM
Screenshot 2026-04-10 at 9 36 22 AM
Screenshot 2026-04-10 at 9 36 16 AM
Screenshot 2026-04-10 at 9 35 37 AM
Screenshot 2026-04-10 at 9 35 27 AM
Screenshot 2026-04-10 at 9 37 16 AM

**Review by robstax — APPROVED** [review]
